File: src/main/java/algs/graph/BreadthFirstSearch.java
package algs.graph;

import algs.shared.util.Testable;

import java.util.LinkedList;

/**
 * Best, Average, Worst: O(V+E)
 */
@SuppressWarnings("unchecked")
public class BreadthFirstSearch implements Testable {

    private final int V;
    private final LinkedList<Integer>[] adj;

    public BreadthFirstSearch() {
        V = 4;
        adj = new LinkedList[V];
        for (int i = 0; i < V; ++i)
            adj[i] = new LinkedList<>();
    }

    public void addEdge(int src, int dest) {
        adj[src].add(dest);
    }

    void bfs(int current) {
        boolean[] visited = new boolean[V];

        LinkedList<Integer> queue = new LinkedList<>();

        visited[current] = true;
        queue.add(current);

        while (!queue.isEmpty()) {
            current = queue.poll();
            System.out.print(current + " ");

            for (int n : adj[current]) {
                if (!visited[n]) {
                    visited[n] = true;
                    queue.add(n);
                }
            }
        }
    }

    @Override
    public void test() {
        addEdge(0, 1);
        addEdge(0, 2);
        addEdge(1, 2);
        addEdge(2, 0);
        addEdge(2, 3);
        addEdge(3, 3);

        bfs(2);

        System.out.println();
    }

}

File: src/main/java/algs/graph/BreadthFirstSearch.kt
package algs.graph

import algs.shared.datastructure.Tree
import algs.shared.util.Testable
import java.util.*


/**
 * BFS
 * https://www.geeksforgeeks.org/breadth-first-search-or-bfs-for-a-graph/
 * Best, Average, Worst: O(V+E)
 */
object BreadthFirstSearch : Testable {

    fun <T> search(value: T, root: Tree<T>): Tree<T>? {
        val queue: Queue<Tree<T>> = ArrayDeque()
        queue.add(root)
        var currentNode: Tree<T>

        while (!queue.isEmpty()) {
            currentNode = queue.remove()
            println("Visited node with value: ${currentNode.value}")
            if (currentNode.value == value) {
                return currentNode
            }
            queue.addAll(currentNode.children)
        }

        return null
    }

    fun <T> search(value: T, start: Node<T>): Node<T>? {
        val queue: Queue<Node<T>> = ArrayDeque()
        queue.add(start)
        var currentNode: Node<T>
        val visited: MutableSet<Node<T>> = HashSet<Node<T>>()

        while (!queue.isEmpty()) {
            currentNode = queue.remove()
            println("Visited node with value: ${currentNode.value}")
            if (currentNode.value == value) {
                return currentNode
            }
            visited.add(currentNode)
            queue.addAll(currentNode.neighbors)
            queue.removeAll(visited)
        }

        return null
    }

    override fun test() {
        testTree()
        testNode()
    }

    private fun testTree() {
        println("Tree Search")

        val root = Tree.of(10);
        val rootFirstChild = root.addChild(2);
        rootFirstChild.addChild(3);
        root.addChild(4);

        search(5, rootFirstChild)
    }

    private fun testNode() {
        println("Node Search")

        val start = Node(10)
        val firstNeighbor = Node(2)
        start.connect(firstNeighbor)

        val firstNeighborNeighbor = Node(3)
        firstNeighbor.connect(firstNeighborNeighbor)
        firstNeighborNeighbor.connect(start)

        val secondNeighbor = Node(4)
        start.connect(secondNeighbor)

        search(5, secondNeighbor)
    }

    @JvmStatic
    fun main(args: Array<String>) {
        test()
    }

}

class Node<T>(val value: T) {
    val neighbors: MutableSet<Node<T>>

    fun connect(node: Node<T>) {
        require(!(this === node)) { "Can't connect node to itself" }
        neighbors.add(node)
        node.neighbors.add(this)
    }

    init {
        neighbors = HashSet()
    }
}
